File: kody_2023/2019_2A/euklides.cpp
/*
 * euklides.cpp
 */

#include <iostream>
using namespace std;

int nwd_1(int a, int b) {
    int i = 0;
    while (a != b) {
        i++;
        if (a > b)
            a = a - b;
        else
            b = b - a;
    }
    cout << "NWD: " << a << endl;
    cout << "Powtórzeń: " << i << endl;
    return a;
}

int nwd_2(int a, int b) {
    int i = 0;
    while (a > 0) {
        i++;
        a = a % b;
        b = b - a;
    }
    cout << "NWD: " << b << endl;
    cout << "Powtórzeń: " << i << endl;
    return b;
}

// NWD(a, b) = b dla a = 0
// NWD(a, b) = { a = a % b; b = b - a; }
// NWD(a, b) = NWD(a % b,  b - a)
int NWD_re1(int a, int b) {
    if (a != 0)
        return NWD_re1(a % b, b - (a % b));
    else
        return b;
}

// NWD(a, b) = a dla b = 0
// NWD(a, b) = {a = b, b = a % b}
// NWD(a, b) = NWD(b, a % b)
int NWD_re2(int a, int b) {
    if (b != 0)
        return NWD_re2(b, a % b);
    else
        return a;
}
// napisz iteracyjną wersję funkcji NWD_re2


int main(int argc, char **argv)
{
    int a, b, i;
    a = b = i = 0;
    cout << "Podaj dwie liczby całkowite: " << endl;
    cin >> a >> b;
    nwd_1(a, b);
    nwd_2(a, b);
    cout << NWD_re1(a, b) << endl;
    cout << NWD_re2(a, b) << endl;
    return 0;
}

File: kody_2025/2AB_inf/euklides.py
def nwd1(a, b):
    licznik = 0
    while a != b:
        licznik += 1
        print(a, b)
        if a > b:
            a -= b
        else:
            b -= a
    print(f'NWD = {a}. Powtórzeń: {licznik}.')

def nwd2(a, b):
    licznik = 0
    while b != 0:
        licznik += 1
        print(a, b)
        # k = b
        # b = a % b
        # a = k
        a, b = b, a % b

    print(f'NWD = {a}. Powtórzeń: {licznik}.')

a = int(input('Podaj liczbę a: '))
b = int(input('Podaj liczbę b: '))

if a > 0 and b > 0:
    # nwd1(a, b)
    nwd2(a, b)
else:
    print('Błędne dane!')
